### PR TITLE
EW-207 Add init job to erwinidm.

### PR DESCRIPTION
**Pull request by HeftyTea**
# Description

Adds dBildungscloud tenant (realm) initialization to erwinidm-role.

## Links to Tickets or other pull requests

- https://ticketsystem.dbildungscloud.de/browse/EW-207

## Changes

Adds initialization job to setup dBildungscloud realm altogether with client and realm specific admin user (to be used by dBildungscloud for administrative access).

## Approval for review

- [ ] All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

> Notice: Please remove the WIP label if the PR is ready to review, otherwise nobody will review it.

### Link to Definiton of Done

More and detailed information on the _definition of done_ can be found [on Confluence](https://docs.schul-cloud.org/pages/viewpage.action?pageId=92831762)

### CODE REVIEW

File: ansible/roles/erwin-idm/templates/init_job.yml.j2 [hunk 1]
```diff
@@ -0,0 +1,32 @@
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: erwinidm-init-job
+  labels:
+    app: api
```
Comment (mamutmk5) [inline ansible/roles/erwin-idm/templates/init_job.yml.j2:6]: api is the wrong label for this job.
```suggestion
    app: erwinidm
```

File: ansible/roles/erwin-idm/templates/init_job.yml.j2 [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+spec:
+  template:
```
Comment (mamutmk5) [inline ansible/roles/erwin-idm/templates/init_job.yml.j2:8]: maybe add an label to the pod for the job
```suggestion
  template:
    metadata:
      labels:
        app: erwinidm-init
```

File: ansible/roles/erwin-idm/README.MD [hunk 3]
```diff
@@ -2,46 +2,59 @@
...
 
-This role will setup ErWIn-IDM and all of its environment. This includes an admin user. No client specific configuration is handled by this role. Please note that this includes the client specific admin user. The configured admin is a global admin. It is not meant to be forwarded to any client application but can be used to create client specific credentials by means of the auto deployment.
+This role will setup ErWIn-IDM and all of its environment. This includes a global admin, which is not meant to be forwarded to any client application but is used to create client specific credentials by means of the auto deployment.
 
-In develop environnement, this role depends on the 'dof_postgresql' role providing persistence. In 'dof_postgresql', a custom 'erwinidm' schema is created during startup. The PostgreSQL role is shared with 'schulcloud-calendar' and 'hydra'.
+This role will setup a dBildungscloud specific realm, client, and corresponding admin user. It does not support partial updates. If the target realm already exists, all further initialization steps are skipped.
```
Comment (mamutmk5) [inline ansible/roles/erwin-idm/README.MD:7]: Don't mention the dBildungscloud at this part.
Write some thing like "This role will setup a instance specific realm, [....]"

File: ansible/group_vars/develop/cfg.yml [hunk 4]
```diff
@@ -2,3 +2,6 @@ FEATURE_MATRIX_MESSENGER_ENABLED: "false"
 FEATURE_INSIGHTS_ENABLED: "false"
 FEATURE_NEXBOARD_ENABLED: "false"
 ROOM_VIEW_ENABLED: "true"
+FEATURE_IDENTITY_MANAGEMENT_ENABLED: "true"
+IDENTITY_MANAGEMENT__TENANT: "dbc"
+IDENTITY_MANAGEMENT__CLIENTID: "dbc"
```
Comment (mamutmk5) [inline ansible/group_vars/develop/cfg.yml:7]: if it's posible for both to go to the namespace as vaule?
```suggestion
IDENTITY_MANAGEMENT__TENANT: "{NAMESPACE}"
IDENTITY_MANAGEMENT__CLIENTID: "{NAMESPACE}"
```

File: ansible/roles/erwin-idm/README.MD [hunk 5]
```diff
@@ -2,46 +2,59 @@
...
-| ERWINIDM_IMAGE_NAME      | The URL to the ErWIn-IDM docker image        |
-| ERWINIDM_IMAGE_TAG       | The ErWin-IDM docker image tag to use        |
-| ERWINIDM_PORT            | Internally user HTTP port for ErWin-IDM      |
-| ERWINIDM_PREFIX          | The ErWin-IDM base url prefix                |
-| ERWINIDM_REPLICAS        | Deployment maximum amount of replica Pods    |
-| ERWINIDM_CPU_LIMITS      | Deployment CPU upper bound                   |
-| ERWINIDM_CPU_REQUESTS    | Deployment CPU lower bound                   |
-| ERWINIDM_MEMORY_LIMITS   | Deployment memory upper bound                |
-| ERWINIDM_MEMORY_REQUESTS | Deployment memory lower bound                |
+Following configuration variables are used for internal and external (dBildungscloud) purpose:
```
Comment (mamutmk5) [inline ansible/roles/erwin-idm/README.MD:29]: Don't mention the dBildungscloud at this part.

File: ansible/roles/erwin-idm/README.MD [hunk 6]
```diff
@@ -2,46 +2,59 @@
...
+
+| Variable                        | Description                                                   |
+| ------------------------------- | ------------------------------------------------------------- |
+| IDENTITY_MANAGEMENT\_\_URI      | The internal base URI of Keycloak                             |
+| IDENTITY_MANAGEMENT\_\_TENANT   | The realm (tenant) name for dBildungscloud                    |
+| IDENTITY_MANAGEMENT\_\_CLIENTID | The default OpenID Connect client of the dBildungscloud realm |
```
Comment (mamutmk5) [inline ansible/roles/erwin-idm/README.MD:35]: Don't mention the dBildungscloud at this part.

File: ansible/roles/erwin-idm/README.MD [hunk 7]
```diff
@@ -2,46 +2,59 @@
...
+| Variable                              | Description                                                |
+| ------------------------------------- | ---------------------------------------------------------- |
+| KEYCLOAK_ADMIN                        | The Keycloak super admin                                   |
+| KEYCLOAK_ADMIN_PASSWORD               | The Keycloak super admin's password                        |
+| KC_DB_URL                             | Connection string to PostgreSQL in JDBC Format [^psqlnote] |
+| KC_DB_USERNAME                        | The PostgreSQL user                                        |
+| KC_DB_PASSWORD                        | The PostgreSQL user's password                             |
+| IDENTITY_MANAGEMENT\_\_ADMIN_USER     | dBildungscloud realm admin                                 |
+| IDENTITY_MANAGEMENT\_\_ADMIN_PASSWORD | dBildungscloud realm admin's password                      |
+| IDENTITY_MANAGEMENT\_\_ADMIN_CLIENTID | dBildungscloud realm admin defined client                  |
```
Comment (mamutmk5) [inline ansible/roles/erwin-idm/README.MD:50]: Don't mention the dBildungscloud at this part.
And sorry that i don't see it the last time

**Review by mamutmk5 — CHANGES_REQUESTED** [review]

File: ansible/roles/erwin-idm/templates/init_job.yml.j2 (hunk 1 above)
Comment (HeftyTea) [reply ansible/roles/erwin-idm/templates/init_job.yml.j2:6]: Done

File: ansible/roles/erwin-idm/templates/init_job.yml.j2 (hunk 2 above)
Comment (HeftyTea) [reply ansible/roles/erwin-idm/templates/init_job.yml.j2:8]: Done

File: ansible/roles/erwin-idm/README.MD (hunk 3 above)
Comment (HeftyTea) [reply ansible/roles/erwin-idm/README.MD:7]: Done

File: ansible/group_vars/develop/cfg.yml (hunk 4 above)
Comment (HeftyTea) [reply ansible/group_vars/develop/cfg.yml:7]: The characters supported for TENANT or CLIENTID are not well documented. The characters [a-zA-Z0-9] are OK. For instance, [+*#?] are invalid. Is NAMESPACE limited somehow? Can we assume no 'special characters' here?

File: ansible/roles/erwin-idm/README.MD (hunk 6 above)
Comment (HeftyTea) [reply ansible/roles/erwin-idm/README.MD:35]: done

File: ansible/roles/erwin-idm/README.MD (hunk 5 above)
Comment (HeftyTea) [reply ansible/roles/erwin-idm/README.MD:29]: done

File: ansible/roles/erwin-idm/templates/configmap-init.yml.j2 [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    IDM_REALM_NAME={{ IDENTITY_MANAGEMENT__TENANT }}
+    IDM_REALM_ADMIN=$IDENTITY_MANAGEMENT__ADMIN_USER
+    IDM_REALM_ADMIN_PWD=$IDENTITY_MANAGEMENT__ADMIN_PASSWORD
+    IDM_REALM_ADMIN_ROLE=realm-admin
+
+    IDM_CLIENT_NAME={{ IDENTITY_MANAGEMENT__CLIENTID }}
+    IDM_CLIENT_URL_BASE=https://{{ DOMAIN }}
+
+    # Wait for IDM to be ready or fail
+    curl --retry 60 --retry-all-errors --retry-delay 10 "$IDM_URL_BASE/health/live" || exit 1
```
[the remark this answers is not in the document]
Comment (HeftyTea) [reply ansible/roles/erwin-idm/templates/configmap-init.yml.j2:28]: That would be more then 13 hours waiting? Ain't that a bit to pessimistic ? 😄 
Maybe 15 minutes will do? Updated that to 90 retries.

[the remark this answers is not in the document]
Comment (mamutmk5) [reply ansible/roles/erwin-idm/templates/configmap-init.yml.j2]: Yes it's 13h but its more about wait til is up there.
That's why it's so an huge amount of time.

File: ansible/group_vars/develop/cfg.yml (hunk 4 above)
Comment (mamutmk5) [reply ansible/group_vars/develop/cfg.yml]: For how the namespace are build see https://github.com/hpi-schul-cloud/dof_app_deploy/blob/main/.github/workflows/deploy.yml#L90
Alternativ you cann posible use SC_THEME for it.
https://github.com/hpi-schul-cloud/dof_app_deploy/blob/main/ansible/group_vars/default/instance_cfg.yml

Comment (HeftyTea) [reply ansible/group_vars/develop/cfg.yml]: Thanks for the hint! SC_THEME seems a good option here.

File: ansible/roles/erwin-idm/templates/configmap-init.yml.j2 (hunk 8 above)
[the remark this answers is not in the document]
Comment (HeftyTea) [reply ansible/roles/erwin-idm/templates/configmap-init.yml.j2]: done

**Review by mamutmk5 — APPROVED** [review]

**Pull request merged**